### ISSUE TITLE
Avoid using setup_requires parameter in setup.py.

### ISSUE THREAD

**Issue by FedericoRessi** [pytest-dev/pytest-xdist#136]
Pip doesn't support installation packages using setup_requires because that option causes distutils to install packages without using PIP. Therefore any PIP configuration are ignored for packages listed in setup_requires. You should instead probably specify project requirements in a file called requirements.txt.

https://pip.readthedocs.io/en/1.4.1/cookbook.html#controlling-setup-requires

**Comment by RonnyPfannschmidt** [comment]
once pip uses https://www.python.org/dev/peps/pep-0518/ (tracked in pypa/pip#3691 )
that won't change unless someone proposes a working solution that is not take out the automation

**Comment by FedericoRessi** [comment]
It is nice to see there this pep 518 that would replace a de-facto unspecified standard with a documented one. Anyway using pip with internal pip mirrors is so common in corporative CI that it is a pity this package doesn't install without configuring the index for easy_install too.

To have the automation without relaying on setup_requires there could be some code in the setup.py that would try to import pip and use it to install as much of those packages as possible before passing remaining list as setup.

Example of setup.py
```python
from setuptools import setup
import traceback

extra_params = {}
setup_requires = ['setuptools_scm']
try:
    import pip
    pip.main(['install'] + setup_requires)
    setup_requires = []
except Exception:
    # Going to use easy_install for
    traceback.print_exc()

setup(
    # [... other parameters here ...]
    setup_requires=setup_requires
    # [... other parameters here ...]
)
```

Using Pip instead of easy_install has a lot of advantages and having to deal with one installation tool instead of two would be nice and make CI behaviour more deterministic.

**Comment by RonnyPfannschmidt** [comment]
@FedericoRessi for stable ci i install tools like pbr/setuptools_scm before installing the main requirements

having crude hack in ci pipelines is deterministic

btw, your example  would mess up a downstream users system state in case they invoked a pip download and hit a sdist (too smart setup.py has been tried, it always breaks someone unexpectedly as well)

**Comment by soerenwolfers** [comment]
This has been fixed in setuptools 42.0.0, which switched to using pip to resolve setup_requires (https://setuptools.readthedocs.io/en/latest/history.html#v42-0-0).
